config/configtls/configtls.go

```diff
@@ -389,17 +394,29 @@ func (c ServerConfig) LoadTLSConfig() (*tls.Config, error) {
 				return nil, err
 			}
 			tlsCfg.GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) { return reloader.getClientConfig(tlsCfg) }
+			c.reloaderShutdownFuncs = append(c.reloaderShutdownFuncs, reloader.shutdown)
 		}
 		tlsCfg.ClientCAs = reloader.certPool
 		tlsCfg.ClientAuth = tls.RequireAndVerifyClientCert
 	}
+
 	return tlsCfg, nil
 }
 
 func (c ServerConfig) loadClientCAFile() (*x509.CertPool, error) {
 	return c.loadCert(c.ClientCAFile)
 }
 
+func (c ServerConfig) Shutdown() error {
+	var err error
+	if c.ReloadClientCAFile {
```

Comment:
Since this value is public, I believe it is possible for it to be set to true when calling `LoadTLSConfig`, which will cause some `shutdowns` funcs to be remembered, then for `ReloadClientCAFile` to be set to false, and then `Shutdown` to be called resulting in the `shutdown` funcs being executed.

I think it would be best to always try to loop through the list and invoke anything inside.

Reply:
Good catch, thanks! I've updated.